test/parallel/test-fetch.mjs

```diff
@@ -0,0 +1,27 @@
+// Flags: --experimental-fetch --no-warnings
+
+import '../common/index.mjs';
+
+import assert from 'assert';
+import events from 'events';
+import http from 'http';
+
+assert.strictEqual(typeof globalThis.fetch, 'function');
+assert.strictEqual(typeof globalThis.Request, 'function');
+assert.strictEqual(typeof globalThis.Response, 'function');
+assert.strictEqual(typeof globalThis.Headers, 'function');
+
+const server = http.createServer((req, res) => res.end('Hello world'));
+server.listen(0);
+await events.once(server, 'listening');
+const port = server.address().port;
+
+const response = await fetch(`http://localhost:${port}`);
+
+assert(response instanceof Response);
+assert.strictEqual(response.status, 200);
+assert.strictEqual(response.statusText, 'OK');
+const body = await response.text();
+assert.strictEqual(body, 'Hello world');
+
+server.close();
```

Comment:
All these lines run very fast, but then on my computer the test hangs a few seconds before it exits. Could it have to do with keep-alive behavior? Can we override it?

Reply:
Try putting a breakpoint here and see if we enter the branch? https://github.com/nodejs/undici/blob/f4f1a862cc4e3c62f9025a9072b7b62b7dc6c471/lib/client.js#L839

Also - is it still connected on the server's side?

Reply:
> Also - is it still connected on the server's side?

How do I know?

Anyway, shouldn't `server.close()` close the connections?

Reply:
> Anyway, shouldn't server.close() close the connections?

https://nodejs.org/dist/latest-v17.x/docs/api/http.html#serverclosecallback

Apparently not.

Reply:
> Try putting a breakpoint here and see if we enter the branch?

Yes, we enter the branch. `client[kKeepAliveTimeoutValue]` becomes `4000`.

Reply:
Can confirm this reproduces in unidici/fetch with a simple tap test I wrote (this takes 4.4 seconds and removing the call reduces it to 400ms)

```js
"use strict";

const { test } = require("tap");

const { fetch } = require("../..");
const { createServer } = require('http')

test("validate fetch doesn't hang the process with keepalive", (t) => {
  t.plan(1)

  const obj = { asd: true }
  const server = createServer((req, res) => {
    res.end(JSON.stringify(obj))
  })
  t.teardown(server.close.bind(server))

  server.listen(0, async () => {
    const body = await fetch(`http://localhost:${server.address().port}`)
    t.strictSame(obj, await body.json())
  })
})
```

Reply:
Looks like the issue is coming from llhttp - when `wasm_on_headers_complete` is called it passes should_keep_alive as true (well 1).

Let me see if I can reproduce on raw llhttp

Reply:
Ok actually let me wait for either @ronag or @indutny to take a look https://github.com/nodejs/llhttp/issues/64 :)

Reply:
ok, apparently I shouldn't debug before I have coffee.

I think llhttp is ok here - undici is sending a keep-alive header and the server is respecting that. The goal is for undici not to send that header.

Reply:
- `undici.fetch` uses the global dispatcher: https://github.com/nodejs/undici/blob/19b5789e2097c44ad1f1279cb8672062abf21030/index.js#L101, which is just an `Agent` created with default options.
- The default value for `keepAliveTimeout` is 4 seconds: https://github.com/nodejs/undici/blob/19b5789e2097c44ad1f1279cb8672062abf21030/docs/api/Client.md#new-clienturl-options

I think this is expected behavior from `undici`'s perspective. The problem is that the `fetch` API doesn't let me change it. I just want to make this test exit as soon as it is finished (without having to use `process.exit(0)` at the end).

Reply:
Workaround:

```diff
diff --git a/test/parallel/test-fetch.mjs b/test/parallel/test-fetch.mjs
index 1aff7e1fdc..de8ef36951 100644
--- a/test/parallel/test-fetch.mjs
+++ b/test/parallel/test-fetch.mjs
@@ -11,7 +11,10 @@ assert.strictEqual(typeof globalThis.Request, 'function');
 assert.strictEqual(typeof globalThis.Response, 'function');
 assert.strictEqual(typeof globalThis.Headers, 'function');
 
-const server = http.createServer((req, res) => res.end('Hello world'));
+const server = http.createServer((req, res) => {
+  res.setHeader('Keep-Alive', 'timeout=0, max=0');
+  res.end('Hello world');
+});
 server.listen(0);
 await events.once(server, 'listening');
 const port = server.address().port;
```

Is this fine?

Reply:
I think users would might find the process 'hanging' for ±4 seconds equally confusing?

Is there a way to eat the cake (use keepalive) and keep it too? (would `.unref`ing the socket when there are no active requests and `ref`ing it when there are work?)

Reply:
I can't answer this.

Reply:
@ronag @Ethan-Arrowood PTAL :)

Reply:
(To be clear - I am totally fine with the workaround and IMO this can be handled after the PR lands)

Reply:
Pushed the workaround.

Reply:
I thought we handled this already. We do unref both timers and sockets.

Reply:
Tragos would you mind making an issue on undici and mention your workaround and we will have a look at it.

Reply:
I think the problem is because the server has an active ref on the socket, not the client.

Reply:
Yeah, I confirmed that this (and removing the keep-alive header) also fixes the issue:

```js
server.on('connection', (socket) => socket.unref());
```

Reply:
The problem for this is not in Undici. We handle it perfectly there.

The problem is Node.js core `http.Server` as the keepalive sockets are... kept alive, preventing Node.js from exiting. In other term, it's https://github.com/nodejs/node/issues/41578.

A solution for this is to set up a different default agent for fetch() that reduces the keepAlive significantly (100ms?). We are not aiming for maximum performance here so it might just be enough to avoid most problems for now.